In [1]:
import torch
from torch import nn
from sklearn import model_selection
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sn
from IPython import display
from sklearn.ensemble import GradientBoostingRegressor
from sklearn.metrics import mean_squared_error, r2_score

import typing
%matplotlib inline

In [2]:
com = pd.read_csv('Top_100_Companies_Load_Balancing_Data.csv')
print(len(com))
#sn.pairplot(diabetes_db)

100


In [3]:
def clean_training_data(df):
    # 1. Remove duplicate rows
    df = df.drop_duplicates()

    # 2. Drop rows with too many missing values (threshold: 50% missing)
    # df = df.dropna(thresh=df.shape[1] * 0.5)

    # 3. Fill remaining missing values with column mean (for numerical data)
    for col in df.select_dtypes(include=['number']).columns:
        df[col].fillna(df[col].mean(), inplace=True)

    # 4. Remove outliers using the IQR method
    Q1 = df.quantile(0.25)
    Q3 = df.quantile(0.75)
    IQR = Q3 - Q1
    df = df[~((df < (Q1 - 1.5 * IQR)) | (df > (Q3 + 1.5 * IQR))).any(axis=1)]

    # 5. Normalize numerical features using StandardScaler
    # scaler = StandardScaler()
    # numeric_cols = df.select_dtypes(include=['number']).columns
    # df[numeric_cols] = scaler.fit_transform(df[numeric_cols])

    # Лексикографическое полуупорядочивание
    df = df.sort_values(by=["Error Rate (%)", "Avg Latency (ms)", "Throughput (Gbps)"], ascending=[True, True, False], kind='stable')
    # Отрезаем 20% последних данных
    df = df.iloc[:round(len(df) * 0.5)]

    return df

In [4]:
for col in com.select_dtypes(include=['object']).columns:
    # factorize returns a tuple (labels, uniques); we only need the labels.
    com[col] = pd.factorize(com[col])[0]
missing_counts = com.isna().sum()
print(missing_counts)
com = clean_training_data(com)
com.fillna(com.mean(numeric_only=True), inplace=True)
com.head(2)

com.drop(columns=['Company'])

Company                       0
Load Balancing Method         0
Users                         0
Servers                       0
Avg Latency (ms)              0
Requests per Second (RPS)     0
Peak RPS                      0
Error Rate (%)                0
Throughput (Gbps)             0
Daily Active Users (DAU)      0
Monthly Active Users (MAU)    0
Avg Session Time (min)        0
Cloud Provider                0
Failover Mechanism            0
dtype: int64


Load Balancing Method       Users  Servers  Avg Latency (ms)  \
25                      5  1280640554    36610             24.17   
12                      3   295108266    51149            129.96   
22                      3  2264382111    53874            149.21   
92                      9  1680638659    12580             98.69   
72                      0   880081176    48127            133.54   
40                      7  2101208734    85902             85.56   
18                      6   192860289     2174            254.19   
86                      7  1776030869    91134            225.66   
2                       2  2871416315    75020            139.58   
27                      1  1833900920     2786            184.12   
6                       5  2768404170    97895            279.44   
84                      6  2868128414    39471            157.02   
21                      2  2791307765    96774             22.14   
39                      6  1068550466    40227             26.09   
74                      2  2419025733    89466            272.58   
99                      4  2851329620    70602             77.47   
75                      3  2805475345    78826             96.72   
38                      7  2802068852     9357             23.30   
37                      7   976488663    27082            181.32   
91                      2  1758575886    72776            100.92   
1                       1   977272515    15595            227.01   
23                      1  2309295631    89787            104.84   
71                      0   233646141    94371             81.20   
98                      5  1734613872    72877             39.36   
97                      7   179980632    30614             16.58   
17                      0   118183524    78693            294.54   
54                      7  2008082416    20773             34.29   
20                      3    68464552     1340            175.61   
70                      1  2693640220    50043            254.24   
51                      3  1154428526    67078            174.62   
49                      0   442642428    75670            221.92   
10                      0   859427829    76829             71.26   
46                      8  2412413623    78082            168.43   
32                      8   528759931    62315            281.54   
5                       4  2354700447    16545             73.93   
93                      2  1618674669    84988            111.47   
4                       3   298711002    90368            208.90   
16                      2  1309729620    99472            216.50   
81                      9   383575777      554            138.50   
14                      6  1883746250    56408            210.63   
64                      9  2378275089    47852             74.77   
13                      6   666294168    98797            110.17   
43                      1  1218494824    60419            278.32   
53                      9   695559762    52472            290.55   
61                      3  2416658063    52781            175.11   
26                      0   406619942    11855            290.46   
0                       0  2336741200     7462            267.58   
90                      1   694898947    85874            224.15   
24                      0   864900382    87475            134.96   
33                      9  2579080829    75794             77.31   

    Requests per Second (RPS)     Peak RPS  Error Rate (%)  Throughput (Gbps)  \
25                    2729373  11273243.71            0.04              59.88   
12                    3802218  14826755.56            0.10             371.76   
22                    1541915   2529604.63            0.14              63.52   
92                    2884495   8558186.74            0.24             451.33   
72                    4239027  13407844.56            0.40             293.37   
40                    4148438  13193582.95       

In [5]:
x_train, x_test, y_train, y_test = model_selection.train_test_split(
    com.loc[:, com.columns != 'Requests per Second (RPS)'],
    com['Requests per Second (RPS)'],
    test_size=0.2,
    random_state=42
    )
x_train = torch.from_numpy(x_train.values).float()
x_test = torch.from_numpy(x_test.values).float()

y_train = torch.from_numpy(y_train.values).float()
y_train = y_train.reshape(-1, 1)

y_test = torch.from_numpy(y_test.values).float()
y_test = y_test.reshape(-1, 1)

print(x_train)

x_train = com.drop(columns=['Requests per Second (RPS)'])
y_train = com['Requests per Second (RPS)']

x_train = torch.from_numpy(x_train.values).float()
y_train = torch.from_numpy(y_train.values).float()
y_train = y_train.reshape(-1, 1)

tensor([[2.1000e+01, 2.0000e+00, 2.7913e+09, 9.6774e+04, 2.2140e+01, 1.8448e+06,
         6.5000e-01, 9.6930e+01, 8.0430e+08, 1.0092e+09, 3.3470e+01, 4.0000e+00,
         1.0000e+00],
        [7.2000e+01, 0.0000e+00, 8.8008e+08, 4.8127e+04, 1.3354e+02, 1.3408e+07,
         4.0000e-01, 2.9337e+02, 2.5867e+08, 7.3409e+08, 2.0280e+01, 2.0000e+00,
         0.0000e+00],
        [1.6000e+01, 2.0000e+00, 1.3097e+09, 9.9472e+04, 2.1650e+02, 1.9764e+06,
         1.8900e+00, 4.3563e+02, 5.1849e+08, 9.6993e+08, 2.7690e+01, 0.0000e+00,
         0.0000e+00],
        [2.0000e+00, 2.0000e+00, 2.8714e+09, 7.5020e+04, 1.3958e+02, 1.2853e+06,
         5.0000e-01, 4.4351e+02, 4.9583e+07, 2.4030e+09, 5.3060e+01, 2.0000e+00,
         2.0000e+00],
        [9.2000e+01, 9.0000e+00, 1.6806e+09, 1.2580e+04, 9.8690e+01, 8.5582e+06,
         2.4000e-01, 4.5133e+02, 6.5916e+08, 1.3241e+09, 3.5410e+01, 4.0000e+00,
         1.0000e+00],
        [1.8000e+01, 6.0000e+00, 1.9286e+08, 2.1740e+03, 2.5419e+02, 3.7741e+06,

In [6]:
gb_regressor = GradientBoostingRegressor(n_estimators=200, learning_rate=0.01, max_depth=10, random_state=42)
gb_regressor.fit(x_train, y_train)

# Делаем предсказание и оцениваем RMSE
preds = gb_regressor.predict(x_test)
# rmse = mean_squared_error(y_test, preds, squared=False)  # RMSE
rmse = mean_squared_error(y_test, preds) ** 0.5  # Используем корень вручную


/opt/anaconda3/lib/python3.12/site-packages/sklearn/ensemble/_gb.py:668: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)  # TODO: Is this still required?


In [7]:
print(y_test.shape)
preds = torch.tensor(preds)
print(preds.shape)

torch.Size([10, 1])
torch.Size([10])


In [41]:
# y_test = torch.tensor(y_test) if not isinstance(y_test, torch.Tensor) else y_test
# preds = torch.tensor(preds) if not isinstance(preds, torch.Tensor) else preds

# Приведение preds к [20, 1] и объединение
result = torch.cat((y_test, preds.view(-1, 1)), dim=1)

In [42]:
result

tensor([[2729373.0000, 2672135.4152],
        [4765297.0000, 4435286.9797],
        [ 343433.0000,  605862.8806],
        [3802218.0000, 3601240.9909],
        [2149987.0000, 2170375.3631]], dtype=torch.float64)

In [43]:
percantage = abs((y_test.squeeze() - preds)/y_test.squeeze() * 100)
print(percantage)
# avg_per = torch.sum(percantage) / percantage.size(0)
# print(avg_per)

tensor([ 2.0971,  6.9253, 76.4137,  5.2858,  0.9483], dtype=torch.float64)


In [44]:
# Фильтрация выбросов > 50%
filtered_percantage = percantage[percantage <= 50]

# Подсчет среднего без выбросов
if filtered_percantage.size(0) > 0:
    avg_per = torch.sum(filtered_percantage) / filtered_percantage.size(0)
else:
    avg_per = torch.tensor(0.0)  # Защита от деления на ноль

print("\nСреднее процентное отклонение без выбросов:")
print(avg_per)


Среднее процентное отклонение без выбросов:
tensor(3.8141, dtype=torch.float64)


Согласно полученным результатам, обученная модель может определять количество запросов в секунду с погрешнотью в 17%. Испытание было проведено на 80 входных данных для тренировки и 20 для тестирования. Использовался алгоритм обучения Gradient Boosting с фильтрацией входных параметров и преобработкой входных параметров (one-hot encoding, удаление строк со слишком большой разницей мат. ожидания)